metals/src/main/scala/scala/meta/metals/providers/DiagnosticsProvider.scala

```diff
@@ -0,0 +1,52 @@
+package scala.meta.metals.providers
+
+import scala.meta.metals.Linter
+import scala.meta.metals.Configuration
+import com.typesafe.scalalogging.LazyLogging
+import scala.{meta => m}
+import scala.meta.metals.ScalametaEnrichments._
+import org.langmeta.lsp.MonixEnrichments._
+import org.langmeta.lsp.PublishDiagnostics
+import org.langmeta.jsonrpc.JsonRpcClient
+import monix.eval.Task
+import monix.execution.Scheduler
+import monix.reactive.Observable
+import org.langmeta.AbsolutePath
+import org.langmeta.lsp.Diagnostic
+
+class DiagnosticsProvider(
+    configuration: Observable[Configuration],
+    cwd: AbsolutePath
+)(implicit s: Scheduler, client: JsonRpcClient)
+    extends LazyLogging {
+  private def latestConfig = configuration.toFunction0()
+  private def scalafixDisabled: Boolean =
+    !latestConfig().scalafix.enabled
+  private def scalacDisabled: Boolean =
+    !latestConfig().scalac.diagnostics.enabled
+
+  lazy val linter = new Linter(configuration, cwd)
+
+  def diagnostics(doc: m.Document): Task[Seq[PublishDiagnostics]] =
+    diagnostics(m.Database(doc :: Nil))
+  def diagnostics(db: m.Database): Task[Seq[PublishDiagnostics]] = {
+    if (scalafixDisabled && scalacDisabled) Task(Nil)
+    else Task.sequence {
+      db.documents.map { document =>
+        val uri = document.input.syntax
+
+        val scalacDiagnostics: Seq[Diagnostic] =
+          if (scalacDisabled) Nil
+          else document.messages.map(_.toLSP("scalac"))
```

Comment:
I was doubting whether this should use the code from `ScalaErrorReporter` instead, like this:

```scala
  private def scalacAnalyze(doc: m.Document): Seq[Diagnostic] = {
    val index = EagerInMemorySemanticdbIndex(
      m.Database(doc :: Nil),
      m.Sourcepath(Nil),
      m.Classpath(Nil)
    )
    for {
      doc <- index.database.documents
      msg <- doc.messages
    } yield msg.toLSP("scalac")
  }
```

I tried it, but it was glitchy so I left it as is. I don't understand very well why `EagerInMemorySemanticdbIndex` was used there and whether it should be used here. Semanticdbs we get here are already built interactively, isn't it enough?